### PR TITLE
Zoom out: Try vertical displacement when dragging a pattern between existing patterns/sections

### PR DESCRIPTION
**Pull request by MaggieCabrera**
## What?

Part of https://github.com/WordPress/gutenberg/issues/50739

Implements a way to separate blocks in zoom out mode vertically both when we drag in a new pattern in between sections and when we click on the inserters

Co-authored-by: Dave Smith <444434+getdave@users.noreply.github.com>

## Why?

We need a better visual aid to guide the user when building with patterns. Separating section makes it clear where the action is happening

## How?

We are adding some conditional separators in between blocks of the blocklist that appear based on a couple of condition: if the block in question is a sectioning element, and if the insertion point is related to said block. The rendering of this separator is animated to push the sections apart.

This also required updating the drag handling logic to avoid triggering `dragend` events when moving in/out of the zoom out separators. This ensures the separator itself does not re-render if the mouse cursor inadvertently moves into/out of it whilst it is animating. This avoids the "stuttering" animations we were seeing early in the lifecycle of this PR.

## Testing Instructions

- Open the inserter and select a pattern category
- On the canvas, click on the plus inserters, they should separate the sections correctly
- Select a pattern and it should insert in the area that was opened
- Try and drag and drop a new pattern in between sections
- The sections should animate and separate while dragging in the area

## Screenshots or screencast 

https://github.com/user-attachments/assets/73d047cb-44d0-4f4e-9134-a9b848566e81

### CODE REVIEW

File: packages/block-editor/src/components/block-list/index.js
Comment (MaggieCabrera) [inline packages/block-editor/src/components/block-list/index.js]: This keeps throwing a `Error: Cannot unlock an object that was not locked before. ` that doesn't make any sense. The code still works so we are ignoring for the time being

File: packages/block-editor/src/components/block-list/content.scss
Comment (MaggieCabrera) [inline packages/block-editor/src/components/block-list/content.scss]: This most likely isn't where this CSS should live. Suggestions welcome! I tried inside block-tools but it seems that CSS is not loaded inside the iframe

File: packages/block-editor/src/components/block-list/index.js
Comment (MaggieCabrera) [inline packages/block-editor/src/components/block-list/index.js]: Same problem we currently have with [the actual inserters](https://github.com/WordPress/gutenberg/blob/6128ceae4fbfcdbb0afbf976c34418a7d5e432ad/packages/block-editor/src/components/block-tools/zoom-out-mode-inserters.js#L34) so we need the same solution for both

File: packages/block-editor/src/components/block-list/index.js
Comment (MaggieCabrera) [inline packages/block-editor/src/components/block-list/index.js]: We could use this same condition to load the div so we don't have a bunch of empty divs in the code, but then we need to figure out a different way to animate them

File: packages/block-editor/src/components/block-list/index.js
Comment (MaggieCabrera) [reply packages/block-editor/src/components/block-list/index.js]: @ellatrix I would love your opinion on this one, cause we can't seem to figure this one out

Comment (MaggieCabrera) [reply packages/block-editor/src/components/block-list/index.js]: @Mamaduka maybe you know?

Comment (Mamaduka) [reply packages/block-editor/src/components/block-list/index.js]: Sorry, I've not seen that error before. But I will try to test it later today or tomorrow.

Comment (MaggieCabrera) [reply packages/block-editor/src/components/block-list/index.js]: I think we fixed it by wrapping it under the zoom out mode condition, but it wasn't really straightforward to figure that out!

**Comment by richtabor** [comment]
I like the exploration; there may be something here. A few notes to see if this can work better: 

1. Let's not have this expand anytime I select a pattern. Only when either the + has been selected, or I'm dragging: 

https://github.com/user-attachments/assets/65756c1a-2458-41af-a6fa-d8906d60ce35

2. There's a discrepancy when I drop in a pattern. I'd expect there to only be one "zone" to drop in a pattern:

https://github.com/user-attachments/assets/a9f5ded1-b983-41a3-b3fc-a9225f256943

3. The existing drop indicator is competing with this zone: 

https://github.com/user-attachments/assets/d5c88311-3cfe-4278-acf6-a1638d215f57

File: packages/block-editor/src/components/block-list/content.scss
Comment (scruffian) [reply packages/block-editor/src/components/block-list/content.scss]: I think this is fine, my only concern is the class name itself, shouldn't it be prefixed with the component?

**Review by scruffian — APPROVED** [review]
This is looking good to me.

**Comment by richtabor** [comment]
We need some sort of delay, or min time: 

https://github.com/user-attachments/assets/e583a46d-bf52-4b51-b995-a1bd7ef28a53

File: packages/block-editor/src/components/block-list/index.js
Comment (Mamaduka) [inline packages/block-editor/src/components/block-list/index.js]: Can we extract this as a separate component? It can be in the same file. Similar render functions are usually hard to maintain.

Comment (scruffian) [reply packages/block-editor/src/components/block-list/index.js]: Good idea. I moved it to a different file as well, as this one is really big.

**Comment by MaggieCabrera** [comment]
> We need some sort of delay, or min time:
> CleanShot.2024-07-26.at.12.28.50.mp4

This is looking like we are allowing the insertion point to be before the header, I don't think that should be the case (at least it isn't on trunk)

File: packages/block-editor/src/components/block-list/zoom-out-separator.js
Comment (Mamaduka) [inline packages/block-editor/src/components/block-list/zoom-out-separator.js]: If we pass `editorMode` from the parent component, we can avoid creating store subscriptions for non-zoomed-out modes.

## How?

If you return early in the `mapSelect` function before calling `select( blockEditorStore )`, no store subscription will be created.

Comment (MaggieCabrera) [reply packages/block-editor/src/components/block-list/zoom-out-separator.js]: Something like [this](https://github.com/WordPress/gutenberg/pull/63896/commits/675d7da09d3ba010d7d98842bda64a919172e2a0)? (and the commit after that...)

Comment (Mamaduka) [reply packages/block-editor/src/components/block-list/zoom-out-separator.js]: Thanks, @MaggieCabrera!

**Comment by MaggieCabrera** [comment]
I don't think this needs a transition delay, though I added one. I see a problem when we drag inside the inserter on top of the first element inside the root element (usually the first item right after the header) where the insertion point keeps being recalculated incorrectly. I haven't figured out yet why. I think once we fix that, this should be good enough to merge

**Review by getdave — COMMENTED** [review]
I love the direction here. I think this kind of interaction will make the experience of drag and drop a lot better.

I appreciate a lot of effort has gone in here and I'd like to acknowledge that before I say that I'm not convinced we have reached the quality bar for this to be shipped. I'll caveat this by noting that it might be we're intending this to intersect with other PRs that resolve some of the issues I've noted here, so if that's the case then please excuse my ignorance as I come up to speed with the project state.

If you take a look the video recording below (including voiceover) you'll see that some drag and drop interactions can lead to unwanted side affects like repeating animations. IMHO we shouldn't be shipping until these issues are ironed out.

It would be beneficial to make a concrete list of the bugs that need to be fixed here and also any interactions that are _missing_ that need to be resolved. I will try and do that (see `Bugs` below).

Thanks for all the effort that's clearly done into this. I know this must be a _really_ difficult problem to solve.

<details>
<summary>Overview video</summary>

https://github.com/user-attachments/assets/b1b07595-035e-467c-82be-167f43ab6699

</details>


<details>
<summary>🐞 Bugs </summary>

### Cycling dropzone animations

We need to fix so that it's not possible to drag and have the dropzones animate underneath the dragged item. 


https://github.com/user-attachments/assets/b1738920-7eec-4bb8-8101-82d047960586

### Multiple dropzone targets

Let's not allow patterns to be dropped within existing patterns. That should happen as a lower level interaction in full editor mode (not zoom out). In fact I wonder if many of these issues will be resolved by fixing this one thing...

### Animation and dropzone below last pattern

I'm not sure we need the full animation with the gray box below the final pattern on the page. Perhaps this is a product of the auto-inserted `core/paragraph` block, but it doesn't feel natural in its current form.


https://github.com/user-attachments/assets/bae9420e-6844-4384-bea3-364922e5fbc4



</details>

**Review by getdave — DISMISSED** [review]
Leaving this blocking review to avoid this shipping until [the issues I noted](https://github.com/WordPress/gutenberg/pull/63896#pullrequestreview-2224250751) are ironed out.

Hopefully merging https://github.com/WordPress/gutenberg/pull/64331 and then rebasing _this_ PR will be the first step.

**Review by getdave — COMMENTED** [review]
https://github.com/user-attachments/assets/751b5d85-538f-4cf3-9cc7-ee7fce7d9e50

I'm also not sure about this interaction when using a keyboard (or assistive tech). We're leaning _very_ heavily on drag and drop and it isn't a great experience.

**Comment by MaggieCabrera** [comment]
Something I noted but I can't investigate further at this moment:

While on the post editor, the root section has a class `has-global-padding` that is causing the separator to not extend the full width. I think we don't need this class present when we are on zoom-out mode.

**Comment by MaggieCabrera** [comment]
Another bug to fix here: when we click on a plus button and then go to select a pattern to insert in the space, the separators collapse. This is because the insertion point changes as soon as the cursor leaves the iframed canvas

**Comment by MaggieCabrera** [comment]
> Another bug to fix here: when we click on a plus button and then go to select a pattern to insert in the space, the separators collapse. This is because the insertion point changes as soon as the cursor leaves the iframed canvas

might be fixed by https://github.com/WordPress/gutenberg/pull/64392

**Comment by richtabor** [comment]
> I'm also not sure about this interaction when using a keyboard (or assistive tech). We're leaning very heavily on drag and drop and it isn't a great experience.

@jeryj where are we on exploring focusing the search field when you select the inverter?

**Comment by getdave** [comment]
@richtabor I'm working on [the Issue](https://github.com/WordPress/gutenberg/issues/63867) in https://github.com/WordPress/gutenberg/pull/64396.

**Comment by getdave** [comment]
## Bugs

- in Site Editor dragging over the header of the template will trigger insertion points outside of the `<main>` element. 
- appearance (show/hide animation) of the zoom out separators causes re-calculation of the blockInsertionPoint which can cause animation stutters or in the worst case trigger opening the separator on a different insertion point

**Comment by getdave** [comment]
Sorry for all the pings. Fixing this now.

**Comment by getdave** [comment]
The main issue with this PR is still that the drop target can change underneath the cursor as a result of the appearance of the larger separator causing the DOM to shift. This causes UI flicker and quirks.

**Comment by richtabor** [comment]
How about this: 

1. Let's not show the blue horizontal line at all. The `block-editor-block-list__zoom-out-separator` space is the indicator, and as-is they're conflicting. 

2. When you "pick up" a pattern, the "space" is added above or below it, while it's faded out in the canvas as well. It would be better if the dragged pattern does not take up space when you're dragging it—it's supposed to be dropped into the space of wherever `block-editor-block-list__zoom-out-separator`, so let that represent it's current placement. 

https://github.com/user-attachments/assets/8770aa4b-0fa7-4dda-90c4-5c7033a8d225

**Comment by MaggieCabrera** [comment]
>  When you "pick up" a pattern, the "space" is added above or below it, while it's faded out in the canvas as well. It would be better if the dragged pattern does not take up space when you're dragging it—it's supposed to be dropped into the space of wherever `block-editor-block-list__zoom-out-separator`, so let that represent it's current placement.
>

What I think needs to happen here is that the "grabbed" pattern needs to disappear from the canvas while it's being dragged so that the effect is much closer to what you'd naturally be expecting

**Review by andrewserong — COMMENTED** [review]
> I love the direction here. I think this kind of interaction will make the experience of drag and drop a lot better.

Same here! I think it makes a huge difference creating room for the pattern that's going to be inserted. I'll just ping @ellatrix for visibility on this PR too, as I know Ella has been looking at ideas for WYSIWYG drag and drop and the displacement here might be interesting in that context.

> We need to fix so that it's not possible to drag and have the dropzones animate underneath the dragged item.

The flickery situation here is a challenge, and I remember we encountered a bit of that in the list view, too, but that situation was a little different. Here's what I'm seeing in testing:

https://github.com/user-attachments/assets/785871b9-6dc1-4a1e-809b-2da47b46b7d6

In terms of what's happening in the editor canvas I'm wondering if there's possibly a couple of things going on:

* Could it be possible that [the throttled function](https://github.com/WordPress/gutenberg/blob/f12b78b7c4e830032d7f9a4641df40236644683b/packages/block-editor/src/components/use-block-drop-zone/index.js#L334) in `useBlockDropZone` is either firing too frequently, or updating when it shouldn't? That said, if I try updating [the throttle to 2000ms](https://github.com/WordPress/gutenberg/blob/f12b78b7c4e830032d7f9a4641df40236644683b/packages/block-editor/src/components/use-block-drop-zone/index.js#L493), I still get the flickery behaviour, which points to:
* While the separator is animating its size, does this result in the editor canvas "thinking" that we've moved the mouse cursor and things need to update? Or, is `showInsertionPoint` being called too frequently?

A couple of possible avenues to investigate further:

* ~See what happens if we greatly increase the throttle~ (I tried that locally and it didn't appear to help, it just made the issue appear more slowly 😅)
* Look into reducing when `showInsertionPoint` is fired while dragging, or only fire it if the insertion point has really changed? Or, within `ZoomOutSeparator` do some additional checks to only update if the insertion point has really changed?

The latter idea could be tricky if the mere presence of the animation results in the insertion point changing 🤔

Another (potentially quite hacky) idea:

* Would it help at all to move the logic for showing/hiding the `ZoomOutSeparator` into a hook so that it's only used in one place, and then use that hook within `block-list/index.js` and then pass down a prop to each of the instances of `ZoomOutSeparator` for showing / hiding the top or bottom separator? In this way, there could be a single hook that is the source of truth for which one to show. The idea here is that then this hook could "own" whether or not to update which one we're displaying, which might allow us to inject some logic that prevents updates when we don't want them to occur (i.e. don't suddenly switch while animating?)

These are just some thoughts / potential rabbitholes to go down in case it helps. Unfortunately it sounds like this will be a tricky bit of behaviour to refine!

**Comment by getdave** [comment]
I've squashed this branch to a single commit to facilitate an easier rebase (there were going to be 18 commits where I'd have to resolve near identical changes). I've now rebased. It seems good to me but pinging @MaggieCabrera for confidence.

**Comment by getdave** [comment]
I noticed that the `<ZoomOutSeparator>` was rendering for `blockInsertionPoint`'s whose operation was other than `insert`.

For example the operation would be `group` and the Zoom Out separator would render even though we'd disabled that ability for drag and drop.

As a result we now have smooth and consistent animations when dragging _within_ a section. 

However, note that this now exposes the real problem which is that when the `<ZoomOutSeparator>` does render it causes the layout to change underneath the cursor thereby triggering the recalculation of the `blockInsertionPoint` thereby causing the `<ZoomOutSeparator>` visibility to change repeatedly.

https://github.com/user-attachments/assets/4df97a27-b96a-42f9-b1be-a0cafcf192d9

**Comment by MaggieCabrera** [comment]
> I noticed that the `<ZoomOutSeparator>` was rendering for `blockInsertionPoint`'s whose operation was other than `insert`.
> 
> For example the operation would be `group` and the Zoom Out separator would render even though we'd disabled that ability for drag and drop.
> 
> As a result we now have smooth and consistent animations when dragging _within_ a section.
> 
> However, note that this now exposes the real problem which is that when the `<ZoomOutSeparator>` does render it causes the layout to change underneath the cursor thereby triggering the recalculation of the `blockInsertionPoint` thereby causing the `<ZoomOutSeparator>` visibility to change repeatedly.

We need to figure out a way to stop recalculating the insertion point when we are hovering the separator

**Comment by getdave** [comment]
In the video below I have set the `<ZoomOutSeparator>` to animate over 3 seconds and also have a red background. If you watch carefully you will see that the animation is interuppted only when the blue insertion line appears. This I believe is a vital clue.


https://github.com/user-attachments/assets/aaebce0c-5002-4cd0-bb79-488f25ab4c96

**Comment by getdave** [comment]
After much sleuthing I narrowed this down to the `onDragLeave` of `useDropZone`:

https://github.com/WordPress/gutenberg/blob/498ce1743750f4a661c96c5883acd6893b165fb0/packages/block-editor/src/components/use-block-drop-zone/index.js#L524-L527

Turns out that `hideInsertionPoint` was being called when moving in/out of a `<ZoomOutSeparator>` because these elements are considered to be outside of the dropzone. 

This was causing the `insertionPoint` to be set to `null` in state which was causing the `getBlockInsertionPoint` selector to change thus causing the re-rendering issues (i.e. both the `ZoomOutSeparator` and the standard "blue line" insertion point to re-render).

~I've used a hacky means to check for the `ZoomOutSeparator`. This proves the issue is what I'm saying above. However, I think we'll need a better way of detecting this.~ (Update: I've reimplemented now)

Check the video to see the "fixed" implementation with no "stuttering":

https://github.com/user-attachments/assets/f2a54856-7f29-4b37-908c-9b00d0e00415

**Comment by MaggieCabrera** [comment]
@getdave this looks so much better now! I don't see anything major beyond the implementation, so it looks like you nailed the problem here. Besides that there are just two more things that need to happen before we merge given [Rich's feedback](https://github.com/WordPress/gutenberg/pull/63896#issuecomment-2289267429):

- We need to remove the blue bar of the insertion point, since the separator is our cue in this mode that we can insert on drag and drop
- When dragging an existing pattern (not a new one) we need to make it visually disappear so the user gets the feedback that the pattern they "grabbed" is moving somewhere else (this might be tricky and mess with insertion points, but maybe it can just be CSS magic?)

**Comment by getdave** [comment]
> @getdave this looks so much better now! I don't see anything major beyond the implementation, so it looks like you nailed the problem here. 

That's great to hear. I'll try and refine the implementation some. It's not great as it is.

> Besides that there are just two more things that need to happen before we merge given [Rich's feedback](https://github.com/WordPress/gutenberg/pull/63896#issuecomment-2289267429):
> 
> * We need to remove the blue bar of the insertion point, since the separator is our cue in this mode that we can insert on drag and drop
> * When dragging an existing pattern (not a new one) we need to make it visually disappear so the user gets the feedback that the pattern they "grabbed" is moving somewhere else (this might be tricky and mess with insertion points, but maybe it can just be CSS magic?)

I would make ~both of those separate PRs~ the 2nd of those a separate PR. I feel _this_ PR will be complicated as it is. It will also make it easier to revert later if we choose to do so for any reason. What do you think?

**Update**: I removed the "blue bar" in https://github.com/WordPress/gutenberg/pull/63896/commits/8ee6d064f55e0d42a75f907af98864fcd5b1f0e4.

File: packages/compose/src/hooks/use-drop-zone/index.js
Comment (getdave) [inline packages/compose/src/hooks/use-drop-zone/index.js]: As this code is within `@wordpress/compose` we cannot refer to "ZoomOut". Therefore need to make this logic abstract such that it can be applied generically to the ZoomOutSeparator.

File: packages/compose/src/hooks/use-drop-zone/index.js
Comment (getdave) [inline packages/compose/src/hooks/use-drop-zone/index.js]: Because this is within the `compose` package we cannot refer to "Zoom out".

Therefore I've aimed to name the concept at work here which is a insertion point between blocks. Hopefully this is suitably generic.

If you apply `data-is-insertion-point` to the element it will be ignored when considering the dragend.

File: packages/block-editor/src/components/use-block-drop-zone/index.js
Comment (getdave) [inline packages/block-editor/src/components/use-block-drop-zone/index.js:444]: There are too many of these conditions creeping in which disallow non-"insert" events in Zoom Out mode. Whilst this is desired behaviour I'd like to codify the concept and have it made reusable. One for a follow up.

File: packages/block-editor/src/components/block-list/index.js
Comment (MaggieCabrera) [inline packages/block-editor/src/components/block-list/index.js:212]: should we use the new [selector](https://github.com/WordPress/gutenberg/pull/64503) here?

Comment (getdave) [reply packages/block-editor/src/components/block-list/index.js:212]: Yes we should! 👍

File: packages/compose/src/hooks/use-drop-zone/index.js
Comment (andrewserong) [reply packages/compose/src/hooks/use-drop-zone/index.js]: Apologies if it's already been explained, but why does this need to happen within the `useDropZone` hook in the `compose` package, rather than within the `onDragLeave` callback in the consumer of the hook? The concept of an insertion point feels bit more conceptually part of the block editor, so looks a bit out of place here to me.

I.e. would it work to perform the insertion point check here instead:

https://github.com/WordPress/gutenberg/blob/f201f8f51d2971969264573327afea0590f82aa0/packages/block-editor/src/components/use-block-drop-zone/index.js#L528-L531

File: packages/block-editor/src/components/use-block-drop-zone/index.js
Comment (andrewserong) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:444]: For now would it work to move this check up to the existing `isZoomOutMode()` check earlier in this code block so that the zoomed out mode checks are consolidated?

Leaving to follow-ups is also a valid approach! My (very) weakly held opinion here is that sometimes it can be tricky to change some of these rules after the fact if we leave them for a while as the drag and drop logic can be a bit challenging to reason about — I know I struggle looking at some of my past code changes to this area of the codebase 😅

**Review by andrewserong — COMMENTED** [review]
Hey, great progress @getdave and @MaggieCabrera, this is feeling really smooth in practical testing! At first I thought I ran into a bug, but it turns out Redux Devtools was causing blocks to throw an error 🤷, so I think it's unrelated to this PR.

Here's how it's looking for me:

https://github.com/user-attachments/assets/c001cdc8-e126-4d8c-839e-067a189314ba

Just left a couple of small code comments — nothing major or blocking, mostly just wondering if we can avoid the change in the `compose` package. But that's more a conceptual concern than an actual problem as no APIs are changing 👍

**Comment by richtabor** [comment]
> When dragging an existing pattern (not a new one) we need to make it visually disappear so the user gets the feedback that the pattern they "grabbed" is moving somewhere else (this might be tricky and mess with insertion points, but maybe it can just be CSS magic?)

Yea, there's tension currently when you grab a pattern to drag it. As the insertion spacer renders above and below this selected pattern (both of which don't actually move the pattern). 

The later half of @andrewserong's video shows this.

File: packages/compose/src/hooks/use-drop-zone/index.js
Comment (getdave) [reply packages/compose/src/hooks/use-drop-zone/index.js]: Good point. I couldn't see this yesterday as I was very fixed on the solution, but revisiting today it seems obvious that it's better to handle in the consumer rather than low level package.

Interesting, this is the very place I was first able to identify and isolate the original bug so I should have known better 😅

File: packages/block-editor/src/components/use-block-drop-zone/index.js
Comment (getdave) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:444]: I'd love to do that but the issue is that `operation` isn't available until this point. That in turn is dependent on other steps above. We could bring all the ZoomOut checks down to _this_ location, but then we'd be doing unnecessary work rather than existing early.

Comment (getdave) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:444]: I feel like we need a code quality PR to track down how many instances of ignoring non "insert" operations in Zoom Out mode exist in the code. Once we know this we can decide whether or not to refactor. It might be premature at this stage if there are only a couple.

Comment (getdave) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:427]: I gave this a go but moving into the consumer causes the fix to break when testing in browser. I can't narrow down why that is.

Comment (getdave) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:427]: Ok I ended up moving the whole thing to the `dragEnd` handler. That has the desired effect.

File: packages/block-editor/src/components/use-block-drop-zone/index.js
Comment (andrewserong) [inline packages/block-editor/src/components/use-block-drop-zone/index.js]: (Very) tiny optimisation idea: this appears to be a pure function. Could it be moved outside of the body of the hook, since it doesn't depend on anything within it?

File: packages/block-editor/src/components/block-list/content.scss
Comment (andrewserong) [inline packages/block-editor/src/components/block-list/content.scss:459]: This probably isn't a huge issue, but I noticed a _very_ subtle sub pixel line of the site's background image poking through while testing:

image

Here it is zoomed in, in case it helps:

image

I imagine this might be because we're scaled down to 50% so the background color here doesn't perfectly extend over the background image?

**Review by andrewserong — COMMENTED** [review]
Code-wise this is looking really close to me, no major issues there from my perspective.

The main issue I ran into is that when testing across browsers, the behaviour appears to be a bit choppy in Safari:

https://github.com/user-attachments/assets/5f499c59-8e3f-419f-a3ce-fda844c4019e

I'm not sure if it's the same issue, but while working on #56625 I ran into an issue with choppiness in Safari that was related to this bug: https://bugs.webkit.org/show_bug.cgi?id=66547. The workaround in https://github.com/WordPress/gutenberg/pull/56625/commits/d9eb4cabc5c7dbbb1f03df71b2055209ece27901 was to add `pointerEvents: none` styling to prevent events from firing as frequently. This might be a long shot, but I wonder if it'd work (or help) to add `pointer-events: none` to the zoomed out separator, since we don't want hovering over the separator to do anything?

From my perspective, since this is behind a feature flag and the scope of the change is fairly contained, once the Safari choppiness is addressed (or verified if it's just a problem I'm running into locally), I reckon if @MaggieCabrera and @getdave are happy with how the PR's looking, it's probably in a good place to merge and continue to iterate on it.

Some things that could be good to continue working on / looking into:

* What @richtabor mentioned above re: the insertion space: https://github.com/WordPress/gutenberg/pull/63896#issuecomment-2306165662
* If you go to drag a pattern that's already in the canvas on a site with a background image, the placeholder state of the pattern you're moving displays as the background image of the site, so it can be a little confusing as to what's happening visually:

https://github.com/user-attachments/assets/675d56c9-acf2-400d-897a-10bc03dcbf37

Nice work continuing to polish this feature, folks!

File: packages/block-editor/src/components/use-block-drop-zone/index.js
Comment (andrewserong) [inline packages/block-editor/src/components/use-block-drop-zone/index.js:554]: I'm not sure of the answer here, so this is just a question! Before the `return;` statement, should we call `throttled.cancel()` to cancel the next cycle of that call, or does it not matter? I'm guessing we're treating this return statement as though nothing happened so _not_ calling `throttled.cancel()` does seem logical to me 👍

File: packages/block-editor/src/components/use-block-drop-zone/index.js
Comment (getdave) [reply packages/block-editor/src/components/use-block-drop-zone/index.js]: Yes you're right. The new location of the function in the consumer will allow this to happen 👍

File: packages/block-editor/src/components/block-list/content.scss
Comment (getdave) [reply packages/block-editor/src/components/block-list/content.scss:459]: Good spot. We'll have to look into that one.

File: packages/block-editor/src/components/use-block-drop-zone/index.js
Comment (getdave) [reply packages/block-editor/src/components/use-block-drop-zone/index.js]: ✅

File: packages/block-editor/src/components/use-block-drop-zone/index.js
Comment (getdave) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:554]: > we're treating this return statement as though nothing happened

That's how I"m seeing it 👍

File: packages/block-editor/src/components/block-list/content.scss
Comment (getdave) [reply packages/block-editor/src/components/block-list/content.scss:459]: ✅ I fixed this by adding an additional 1px

**Comment by getdave** [comment]
Rebased to pull in fix for Safari bug to allow testing on older versions of Safari.

**Comment by getdave** [comment]
In relation to the jankyness of the separator animation in Safari. This is because the `relatedTarget` is not set when the `dragleave` event fires when leaving a block and entering the separator (caused by the animation).

The info on the draglave targets is as follows:

- `target` - The [EventTarget](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget) the pointing device exited from
- `relatedTarget` - The [EventTarget](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget) the pointing device entered to

As the `relatedTarget` is `null` we cannot detect that the drag has started over a separator (like we can in Chrome) an thus we cannot cancel the `dragLeave` handler on `useBlockDropZone`. This means that the `hideInsertionPoint` action is dispatched which causes the flicker.

The bug in Safari is - https://bugs.webkit.org/show_bug.cgi?id=66547

We now need to find a workaround.

File: packages/block-editor/src/components/block-list/content.scss
Comment (andrewserong) [reply packages/block-editor/src/components/block-list/content.scss:459]: Nice idea 👍

**Comment by andrewserong** [comment]
Thanks for digging into the jankiness @getdave! Just re-testing `trunk` on Safari, I notice there's some jankiness there, too, in the left and right areas of the canvas area in zoom out mode, with areas where the drag chip gets stuck:

https://github.com/user-attachments/assets/caf9b9bd-60ca-4c5d-b7c8-8030504d9b89

So, I'm wondering if the source of the jankiness issue is partly something that already exists in `trunk`, but is more visible in this PR due to the animated separators, whereas the drop indicator lines are bit more resilient to hiding jankiness since there isn't any animation.

I'm a bit on the fence about whether or not the jankiness is a blocker to landing — it'd be a great thing to fix and polish, but I also don't want to hold up this PR. Do feel free to go ahead and land this if you and others are happy enough with the current state of the PR!

**Comment by MaggieCabrera** [comment]
My vote is YES to land this and polish after. This is a much better experience than trunk already

**Review by scruffian — APPROVED** [review]
This is looking great, thanks for all the hard work

**Comment by scruffian** [comment]
I added an issue to track the ongoing problems in Safari: https://github.com/WordPress/gutenberg/issues/65125

**Pull request merged**
